Draft of Upstream Documentation ADR (#7)
operate-first/SRE: ADRs/RH/SIG-SRE/ADR-00002 Open Source Documentation.md

## Goals

* Propose a model for maintaining 2 versions of documents, for upstream and downstream consumption
* Provide a worked example of this model as a reference

## Non-goals

* Implement a complete solution for a product or service

## Current Architectures

A Standard Operating Procedure (SOP) is a document used by SREs in response to alerts and other events.
An example of what a current SOP architecture looks like is the Red Hat OpenShift Streams for Apache Kafka (RHOSAK) managed service. The RHOSAK SOPs are located in a private repository, as the SOPs contain references to Red Hat specific internal information that is not suitable for consumption by external users. Such information includes links to Vault secrets, information about how RHOSAK specifically set up their stage and production environments and information on the infrastructure itself, for example what certain custom resources should look like. The flow of how information is contributed to the RHOSAK SOPs is for an engineer or SRE to submit PRs to the repo.

## Proposed Architecture

> The proposed solution is to follow a two repository model. One private repository that will contain SOPs that have the Red Hat internal information mentioned above, the other being a public repository that will have SOPs with the Red Hat internal information removed. This model will allow the SRE teams to use the private SOP repository without any information loss and allow upstream users to use and contribute to the SOP without breaching any Red Hat data policies.

Reviewer: Some comments from implementing this for the RHEL/Cyborg docs:
- it works very nicely to have a public repo with all the docs, and an internal one with only the confidential bits, and *no mirroring* (because there is a publishing step involved); the merging/syncing will be done during publishing, when the upstream and downstream versions of the documentation are built
- it was commented during review that we needed to provide some hints whether somebody is looking at the upstream or downstream documentation, we selected the main theme color (blue <-> red) to distinguish between https://cyborg-infra.gitlab.io/documentation/general/contributing/ and https://cyborg.pages.redhat.com/documentation/general/contributing/
- if there are two versions of a page (one without and one with the confidential bits), it is nice to be able to jump between those versions jic you are looking at the wrong one; we selected a banner with either "This page has an *internal companion page* which might contain additional information." or something like "Some of this information is meant for internal use only. Do not share this information in public unless it is also included on the *public companion page*."

> The proposed solution is to follow a two repository model. One private repository that will contain SOPs that have the Red Hat internal information mentioned above, the other being a public repository that will have SOPs with the Red Hat internal information removed. This model will allow the SRE teams to use the private SOP repository without any information loss and allow upstream users to use and contribute to the SOP without breaching any Red Hat data policies.

Reviewer 2: Thank you for proposing this architecture! Have we already considered an overall logical split of documentation instead of filtered mirroring? 

For example, having an internal/private 'runbook' for incidents, with links to external SOP docs. In the case of many upstream projects, there's an ongoing effort to publish runbooks for the alerts that ship with the project. Here's a good one: https://runbooks.prometheus-operator.dev

So in this case, we could look at it this way:
* The 'internal' runbook would have information on specifics for our implementation of the upstream project, access details, configuration documentation etc. 
* The 'upstream' SOP's will be used exclusively for troubleshooting and resolution steps. 

This way, we would adhere to an 'upstream-first' principle and contribute directly to the upstream repositories instead of having to do a manual filter/processing of the runbooks. 

What do you think?

> The proposed solution is to follow a two repository model. One private repository that will contain SOPs that have the Red Hat internal information mentioned above, the other being a public repository that will have SOPs with the Red Hat internal information removed. This model will allow the SRE teams to use the private SOP repository without any information loss and allow upstream users to use and contribute to the SOP without breaching any Red Hat data policies.
> 
> ![Upstream, Downstream Pattern](images/upstream_downstream_docs_pattern.png)
> 
> The content of the SOPs will be mirrored/synced between both repositories apart from the Red Hat internal information. The flow of contributing information to the repositories can be done both ways i.e. from downstream to upstream and from upstream to downstream. This information can be contributed manually through the creation of pull requests or it can be done autonomously using the mirroring features in the repository. The team in charge of the SOPs will be responsible for maintaining the information between both repositories. This includes merging contributions from external users in the upstream repository and ensuring, if applicable, it gets added to the downstream repository and vice versa.
> 
> SOPs from a private downstream repository that contain sensitive information, will have to be cleaned of all sensitive information before being pushed to the upstream public repository. Such information might be links to vault secrets or information about the managed service infrastructure etc. The pre-existing Structure of  the SOPs should also be looked at, to see if any information can be removed that is no longer necessary or can be moved to a different internal location, this will help reduce overhead when syncing.

Reviewer: From the [Open Source requirements for Services]:
> - A public contributor can use the same workflow as a typical team member to make a change to the service.

If a "change to the service" would also include changes to its documentation or runbooks, this would mean that the "main" documentation repository should be the upstream one. The downstream one should then only be used for changes of internal details.

[Open Source requirements for Services]: https://www.operate-first.cloud/community/open-source-services.html

```suggestion
The proposed solution is to follow a two repository model. The public repository contains the SOPs without any company internal information, and is the main documentation repository. The internal repository contains (sensitive) company internal information.

For changes that only contain public information, only one change request to the public repository is needed. For changes that also contain internal information, two change requests are needed: one to the public repository with the public information, and one to the internal repository with the internal information. To enable successful merging of the information from both repositories, public and internal information should not be intermingled in any one file.

![Upstream, Downstream Pattern](images/upstream_downstream_docs_pattern.png)

SRE teams need to be able to use SOPs with the internal information included. With the proposed repository layout, this could be implemented in multiple ways, e.g.:
- all changes to the public repository are synchronized to the internal repository, so the internal repository by itself will contain the SOPs with internal information included
- during a publishing step, two versions of of the final documentation are generated: (1) a public (upstream) version, only containing information from the public repository, and (2) an internal (downstream) version, containing information from both the public and internal repositories
```

Author: > If a "change to the service" would also include changes to its documentation or runbooks, this would mean that the "main" documentation repository should be the upstream one. The downstream one should then only be used for changes of internal details.

@mh21 agreed.
After careful review of the feedback here & in the gdoc internally that this ADR is based on, I'm going to close this PR and open a new one that takes a more risk averse approach that should match what you're saying.
(The internal link for the new ADR is https://docs.google.com/document/d/11o88BcDLAGOHCl3Lo3F7OM6bvQLhDvH7gUreVIImnq8/edit , which I intend to make a generic version of for here)

# ADR-00002 Open Source Documentation

Authors: Rachel Lawton, David Martin

## Status

Draft.

## Problem Statement

We want to build an upstream community around managed services. This means all aspects of the managed service including the ability to run and manage the service, which extends to operational knowledge & procedures. One such element is Standard Operating Procedures (SOPs). These SOPs include information like how to scale up the service as load increases as well as how to diagnose and fix specific issues based on alerting. However, currently these SOPs have been written from an internal perspective and are not immediately usable or consumable upstream.

> This document proposes a solution for maintaining an upstream version of the SOPs and having  a mechanism for consuming them downstream, as well as vice versa.

Reviewer: ```suggestion
This document proposes a solution for maintaining an upstream version of the SOPs and having a mechanism for consuming them downstream, as well as vice versa.
```

## Goals

* Propose a model for maintaining 2 versions of documents, for upstream and downstream consumption
* Provide a worked example of this model as a reference

## Non-goals

* Implement a complete solution for a product or service

> ## Current Architecture

Reviewer: ```suggestion
## Current Architectures
```

A Standard Operating Procedure (SOP) is a document used by SREs in response to alerts and other events.
An example of what a current SOP architecture looks like is the Red Hat OpenShift Streams for Apache Kafka (RHOSAK) managed service. The RHOSAK SOPs are located in a private repository, as the SOPs contain references to Red Hat specific internal information that is not suitable for consumption by external users. Such information includes links to Vault secrets, information about how RHOSAK specifically set up their stage and production environments and information on the infrastructure itself, for example what certain custom resources should look like. The flow of how information is contributed to the RHOSAK SOPs is for an engineer or SRE to submit PRs to the repo.

> 

Reviewer: Another current example would be the CKI or image builder docs as linked in another comment, which basically has SOPs split across the two repositories, but never presenting *all of them* in a single place.

Reviewer: This is now deployed for the RHEL/Cyborg team documentation: one public repo with the docs, one internal repo with the confidential bits, and during deployment/review those pieces get merged automatically: https://cyborg.pages.redhat.com/documentation/general/contributing/

## Proposed Architecture

The proposed solution is to follow a two repository model. One private repository that will contain SOPs that have the Red Hat internal information mentioned above, the other being a public repository that will have SOPs with the Red Hat internal information removed. This model will allow the SRE teams to use the private SOP repository without any information loss and allow upstream users to use and contribute to the SOP without breaching any Red Hat data policies.

![Upstream, Downstream Pattern](images/upstream_downstream_docs_pattern.png)

> The content of the SOPs will be mirrored/synced between both repositories apart from the Red Hat internal information. The flow of contributing information to the repositories can be done both ways i.e from downstream to upstream and from upstream to downstream. This information can be contributed manually through the creation of pull requests or it can be done autonomously using the mirroring features in the repository. The team incharge of the SOPs will be responsible for maintaining the information between both repositories. This includes merging contributions from external users in the upstream repository and ensuring, if applicable, it gets added to the downstream repository and vice versa.

Reviewer: ```suggestion
The content of the SOPs will be mirrored/synced between both repositories apart from the Red Hat internal information. The flow of contributing information to the repositories can be done both ways i.e. from downstream to upstream and from upstream to downstream. This information can be contributed manually through the creation of pull requests or it can be done autonomously using the mirroring features in the repository. The team in charge of the SOPs will be responsible for maintaining the information between both repositories. This includes merging contributions from external users in the upstream repository and ensuring, if applicable, it gets added to the downstream repository and vice versa.
```

> The content of the SOPs will be mirrored/synced between both repositories apart from the Red Hat internal information. The flow of contributing information to the repositories can be done both ways i.e from downstream to upstream and from upstream to downstream. This information can be contributed manually through the creation of pull requests or it can be done autonomously using the mirroring features in the repository. The team incharge of the SOPs will be responsible for maintaining the information between both repositories. This includes merging contributions from external users in the upstream repository and ensuring, if applicable, it gets added to the downstream repository and vice versa.

Reviewer: I'm not sure how automatic mirroring downstream->upstream would interact with the idea of having an open-source workflow for the SOPs. It would mean that the primary source of content will be the internal downstream repo, which will prevent external contributions.

Is the main reason for the duplication the idea to have all SOPs in one place? Afaict SOPs are linked e.g. in alerts, so whether they live in the public or private shouldn't matter much 🤔. So it should be possible to have SOPs that only live either in the upstream or in the downstream repo.

All documentation I know of is rendered via hugo/mkdocs etc. So if it is still desirable to have all SOPs in one place, maybe there would be a way to merge documentation while rendering. Something we have been experimenting with is to have the SOPs public without any sensitive information like hostnames etc, and then provide a link to a separate internal page that lists those sensitive details. Another idea might be to mess with the rendering in a more sophisticated way:
- external upstream repo with readable SOPs that don't contain any sensitive info -> renders nicely for public consumption
- internal downstream repo with internal details only -> pulls in external upstream repo with SOPs (submodule/git-pull) -> rendering merges the internal details into the SOPs (by include/script ✨magic✨ etc) -> renders nicely for internal consumption

EDIT: reading some more, that is more or less the approach described in [ADR 84](https://docs.google.com/document/d/1qVtEWm71snegFPT3IccShRGV6YFdxMwE7AXjPY0F10M) 😂

We have a page that we want to move to the Cyborg docs about onboarding AWS accounts (https://documentation.internal.cki-project.org/infrastructure/aws-infrastructure/), and that could be used as a guinea pig for such a setup (i.e. have it available in both places automatically, but in the internal place also somehow inject the sensitive information). Let me know whether that seems something worth to try in the scope of this ADR.

Author: > reading some more, that is more or less the approach described in [ADR 84](https://docs.google.com/document/d/1qVtEWm71snegFPT3IccShRGV6YFdxMwE7AXjPY0F10M)

@mh21 Yes, exactly. This PR is a draft upstream version of that ADR (practically the same but with some RHOSAK specifics omitted). A lot of comments have been made on the ADR, which can feed back into changes here when they conclude

> SOPs from a private downstream repository that contain sensitive information, will have to be cleaned of all sensitive information before being pushed to the upstream public repository. Such information might be links to vault secrets or information about the managed service infrastructure etc. The pre-existing Structure of  the SOPs should also be looked at, to see if any information can be removed that is no longer necessary or can be moved to a different internal location, this will help reduce overhead when syncing.

Reviewer: ```suggestion
SOPs from a private downstream repository that contain sensitive information, will have to be cleaned of all sensitive information before being pushed to the upstream public repository. Such information might be links to vault secrets or information about the managed service infrastructure etc. The pre-existing Structure of the SOPs should also be looked at, to see if any information can be removed that is no longer necessary or can be moved to a different internal location, this will help reduce overhead when syncing.
```
